src/box/tuple_compare.cc

```diff
@@ -105,6 +105,29 @@ mp_extension_class(const char *data)
 	return mp_ext_classes[type];
 }
 
+/**
+ * Reads a number from the MsgPack @p field and casts it to double.
+ */
+static double
+mp_read_as_double(const char *field)
```

Comment:
Hash and tree indexes behave differently now. I'm not sure this is a problem though. @alyapunov.

<details>

```
tarantool> _ = box.schema.space.create('tree')
---
...

tarantool> _ = box.space.tree:create_index('pk', {type = 'tree', parts = {1, 'double'}})
---
...

tarantool> _ = box.schema.space.create('hash')
---
...

tarantool> _ = box.space.hash:create_index('pk', {type = 'hash', parts = {1, 'double'}})
---
...

tarantool> x = 1e63-1ULL
---
...

tarantool> box.space.tree:insert{x}
---
- [18446744073709551615]
...

tarantool> box.space.tree:insert{tonumber(x)}
---
- error: Duplicate key exists in unique index "pk" in space "tree" with old tuple
    - [18446744073709551615] and new tuple - [1.84467e+19]
...

tarantool> box.space.hash:insert{x}
---
- [18446744073709551615]
...

tarantool> box.space.hash:insert{tonumber(x)}
---
- [1.844674407371e+19]
...
```

</details.

Reply:
Fixed, now double keys are hashed as appropriate: first we cast the value in the field to double and then we make the double msgpack out of it and hash it.